### PR TITLE
Prototype

### PR DESCRIPTION
**Pull request by tiboloic**
First prototype of proemis3D, as demonstrated in Melbourne

Implements most functionalities except:

- the Alignment component 
- the o/e model of gnomAD v2.1

Known problems that do not need to be pointed out:

- all file names are hard-coded, will be fixed later
- mostly declarative style, will be properly encapsulated in functions later
- Roxygen2 style comments still need to be added to clarify the arguments and return values of functions

### CODE REVIEW

File: src/3DFactory.R [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+## calculate upper bound of the o/e confidence interval
+get_oe_ci_up <- function(observed, expected, alpha = 0.05) {
+  return(qchisq((1 - alpha / 2), 2 * (observed + 1)) / 2 / expected)
+}
+
+## calculate intolerance metric of a region
+get_region_metric <- function(i_region, alpha, regions, oe_data) {
+  # indices of the residues belonging to this region
+  this_region <- regions[seq_len(i_region)]
```
Comment (KatalinaBobowik) [inline src/3DFactory.R:28]: What do you mean by ‘this_region’? I’d change this to be more descriptive

File: src/3DFactory.R [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+# 3D Factory
+
+# get 3D region and its metrics for each residue
+regions_list <- lapply(seq_len(n), function(i_residue) {
+
+  # get the corresponding row of the distance matrix
+  row <- dist_mat[i_residue, ]
+
+  # grow the sphere = order by increasing distance
+  this_aa_regions <- seq_len(n)[order(row)]
```
Comment (KatalinaBobowik) [inline src/3DFactory.R:71]: Again, I’d  choose a variable name without ‘this’

File: src/3DFactory.R [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+# Adjudicator
+
+# creates a partition of the protein into regions of intolerance
+# assigns to each residue a single intolerance score
+# prototype: greedy algorithm: discussion with Sabrina and Leo, 28/04/2022
+
+## sort list of regions by increasing intolerance score and create map of residue to region
+
+# intolerance scores
+oes <- sapply(regions_list, function(one_region) one_region$metric["oe"])
```
Comment (KatalinaBobowik) [inline src/3DFactory.R:98]: Out of curiosity, what does ‘oes’ stand for here? I’m not sure if other people will be confused, or if it’s just because I’m not in the field

File: src/3DFactory.R [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+for (i_region in seq_len(n)) {
+  this_region <- regions_list[[i_region]]
+
+  # residues in this region that do not have an intolerance score yet
+  to_score <- which(is.na(residue_to_region[this_region$region, "region_index"]))
+
+  # assign the regions's intolerance score to these residues
+  residue_to_region[this_region$region, "region_index"][to_score] <- i_region
+
+  # end if all residues have been scored
```
Comment (KatalinaBobowik) [inline src/3DFactory.R:116]: cool 👌

**Review by KatalinaBobowik — APPROVED** [review]
Super clear, great job Loic. Just added a couple nit comments

File: src/3DFactory.R (hunk 3 above)
Comment (tiboloic) [reply src/3DFactory.R:98]: an old habit: when I have a list or set or vector of stuff I often just add an s at the end to make it clear in the variable name that there are multiple elements. So 'oes' = a list of oe. 'oe' is constraint jargon standing for 'observed over expected ratio' which directly estimates the proportional depletion in mutations for a genomic region. That's the oe that appears in 'LOEUF' the loss of function intolerance metric developed by Konrad Karczewski and also the 'oe' in prOEmis3d.

File: src/3DFactory.R [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+
+  # get the corresponding row of the distance matrix
+  row <- dist_mat[i_residue, ]
+
+  # grow the sphere = order by increasing distance
+  this_aa_regions <- seq_len(n)[order(row)]
+
+  # calculate the upper bound of the o/e confidence interval for each region
+  regions_metric <- sapply(seq_len(n), function(i) {
+    get_region_metric(i, 0.2, this_aa_regions, oe_data = oe)
```
Comment (katiedelange) [inline src/3DFactory.R:75]: What's the reasoning behind alpha = 0.2? 

On an initial read through that seems very generous, but I assume there's good logic behind it - so just to note that, while I know adding the full roxygen2 documentation to the function definition is already on your radar, wherever possible if you can also include explanations for appropriate default parameter values as part of that documentation that would be great.

File: src/3DFactory.R (hunk 1 above)
Comment (katiedelange) [inline src/3DFactory.R:28]: Not a blocker on this PR, but it would be interesting to think about how you might expand the region creation step to test spheres that can't be built through this iterative add-the-nearest-residue approach. 

I.e. as long as you do this exhaustive search (adding a single residue at a time, testing every single residue as the starting point for region formation), it should be great for capturing conserved regions if they are centred on a given residue. But for things like binding pockets, you could imagine the centre point for the conservation signal would actually be _between_ residues. Forming spheres around any of the relevant pocket residues may still capture signal, but it's likely to be diluted by non-binding-pocket residues that are in more immediate proximity to the starting point.

**Review by katiedelange — APPROVED** [review]
Looks great: nicely written and easy to follow. Looking forward to the additional roxygen2 documentation when you get a chance.

One small nitpick: whenever possible, it would be great if you could make a separate PR to merge any formatting changes (which are then super easy to approve), as well as a separate PR for large-scale additions of test data. Small, specific changes to test data should be kept with the code changes that required them. This then makes it much easier to pick out the actual code changes that need review and make sure nothing gets lost.

File: src/3DFactory.R (hunk 5 above)
Comment (tiboloic) [reply src/3DFactory.R:75]: The easy and not satisfying answer to this question is that alpha = 0.2 is not far from the value used by Konrad Karczewski for LOEUF
In truth it is largely arbitrary. What alpha controls is the [trade-off between bias and variance](https://towardsdatascience.com/understanding-the-bias-variance-tradeoff-165e6942b229). Low values of alpha will favour bias over variance (i.e it will generate fewer regions of larger size, with the risk of underfitting), larger values of alpha will favour variance over bias, with the risk of overfitting. Here I picked 0.2 arbitrarily because I was worried of under-fitting and kept it because it gave decent results for the demo.
In the future my plan is to use a proper model selection procedure to choose the 'best' value for alpha. Most likely I will estimate alpha using maximum likelihood , but not completely sure yet because the statistical model needs to be refined: a simple Poisson model for the observed value  would underestimate the variability by ignoring the uncertainty on the expected

File: src/3DFactory.R (hunk 1 above)
Comment (tiboloic) [reply src/3DFactory.R:28]: Yes it is on my list. Technically it is simple to make the same calculations using an arbitrary centre for the sphere: just use the 3d coordinates of the alpha-carbons and calculate the distance between the centre of the sphere and each residues. It is computationally cheap.
The real question therefore is how to choose the position of the centre of the sphere. I have so far considered three strategies:
- brute force. Generate a large number of spheres with random positions, keep the best ones. Computationally expensive.
- optimization. Starting from a candidate sphere centre, optimize its position using gradient descent. This would require the intolerance metric to be continuous and derivable over space which could be achieved, for example, by using a Gaussian kernel
- weighted average / barycentre. Find intolerant spheres close to each other and use the barycentre of their centres to define a new candidate sphere

All this needs further thought but before complicating the approach I want to collect hard evidence that it is needed. This is why we have chosen a couple of proteins with constrained beta-barrels (a good model were you could expect the best sphere to have its centre in the middle of the barrel, away from any residue) for the validation phase

Comment (tiboloic) [reply src/3DFactory.R:28]: Regarding [this comment](https://github.com/populationgenomics/structural-constraint/pull/28#discussion_r884199637)
Agreed, this was  temporary for the demo.
Regarding formatting: we made the decision of using `styler` after I had started coding. In the future the formatting should be right from the beginning, especially when we have the pre-commit hooks
Regarding test data: my view is that the results of analyses should not even be on Github, or if they are they should be in a separate directory, but am not sure how to exactly handle that. Suggestions welcome

Comment (tiboloic) [reply src/3DFactory.R:28]: Agreed. Changed to `region_indices`

File: src/3DFactory.R (hunk 2 above)
Comment (tiboloic) [reply src/3DFactory.R:71]: Agreed. Changed to `current_sphere_regions`

File: src/constraint_utils.R [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+get_dist_mat <- function(pdb_file) {
+  pdb <- read.pdb(pdb_file)
+
+  # calculate distance between residues based on their carbon alpha
+  dist_mat <- dm(pdb, inds = "calpha")
+
+  # dist_mat is upper triangular with empty diagonal: add diagonal, fill lower triangle, convert to matrix
+  dist_mat <- as.matrix(dist_mat)
+  dist_mat[is.na(dist_mat)] <- 0
+  dist_mat <- dist_mat + t(dist_mat)
```
Comment (silkm) [inline src/constraint_utils.R:22]: Think this simplifies this a little. I don't think you need to specify `as.matrix`, it should still be class `dmat`

```suggestion
  dist_mat <- dm(pdb, inds = "calpha", mask.lower = FALSE)
  dist_mat[is.na(dist_mat)] <- 0
```

File: src/constraint_utils.R [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+## iCn3D
+
+export_iCn3D <- function(oe_data, file) {
+  if ("numeric" %in% class(oe_data)) {
+    oe <- oe_data
+  } else {
+    oe <- oe_data[, "oe"]
+  }
+
+  # set score on 0-100 range (needed by iCn3D)
```
Comment (silkm) [inline src/constraint_utils.R:37]: just a comment, nothing to action:
re: allowing 0-100 and no NA values, I think this is also a quirk with the other protein viewers as well that we might need to factor in later, though potentially more annoying if we can't use negative numbers.

So for residues we don't want coloured we can normally set the score as -1, then in a viewer select residues with -1 and colour something neutral. If we can't use negatives we might need to handle it differently

File: src/3DFactory.R (hunk 1 above)
Comment (silkm) [reply src/3DFactory.R:28]: Or potentially just `region < regions[ ... ]` showing it's a subset? `region_indices` could replace `i_region`

**Review by silkm — APPROVED** [review]
Looks great, really clear. Added a few comments but nothing needing changes for the PR or current prototype. Especially love the data structure approach in the 3D factory
